# Løse likninger med `sympy`

Å løse likninger er en viktig del av matematikken. I dette notatet skal vi se på hvordan vi kan løse likninger med `sympy`.

## Likninger med én ukjent

For å løse likninger, bruker vi funksjonen `sympy.solve`. Denne funksjonen tar to argumenter,

1. Venstesiden av likningen som skal løses. Det er antatt at likningen er på formen $f(x) = 0$, så vi skal bare sende inn hva $f(x)$ er.
2. Den ukjente variabelen som skal løses for.

Den generelle skrivemåten blir derfor

```python
import sympy # Må importere biblioteket
from sympy.abc import x # Importerer en symbolsk variabel for `x`

løsninger = sympy.solve(funksjonsuttrykk) # Løser likningen f(x) = 0 for x.
```


````{admonition} Andre måter å lage en symbolsk variabel på 
:class: tip

Legg merke til at vi må importere både `sympy` som et bibliotek, samt `x` som en symbolsk variabel fra `sympy.abc`. 
Python vil ikke forstå hva vi prøver å løse hvis vi ikke importerer `x` som en symbolsk variabel fra `sympy.abc`, 
og vi vil bare få at `x` ikke er definert. 

Alternativt kan `x` hentes ut som en symbolsk variabel fra `sympy` på følgende måte:

```python
x = sympy.abc.x
```

````


### Eksempel 1: Løse en lineær likning

La oss se på hvordan vi kan løse likningen

$$
2x - 1 = 0.
$$

Dette kan vi selvfølgelig gjøre for hånd, og da finner vi at $x = 1/2$. 

For å bruke `sympy` må vi identifisere hva $f(x)$ er for noe, og i dette tilfellet er det $f(x) = 2x - 1$. Vi kan derfor skrive

In [40]:
import sympy
from sympy.abc import x, y # Lager en symbolsk variabel for x

f = 2 * x - 1 # Lager en symbolsk funksjon f(x) = 2*x - 1
løsninger = sympy.solve(f) # Løser likningen 2*x - 1 = 0.

print(løsninger)

[1/2]


som betyr at løsningen av likningen er $x = 1/2$.

````{admonition} Kommentar til koden
:class: tip, dropdown

I eksempelet her, og i de neste eksemplene som følger, har definert en symbolsk uttrykk for funksjonen `f` i likningen $f(x) = 0$.
Man trenger strengt tatt ikke å gjøre dette, men kan i stedet bare skrive venstresiden av likningen direkte inn i `sympy.solve`:

```python
løsninger = sympy.solve(2 * x - 1)
```

Dette vil gi akkurat samme resultat.

````

### Eksempel 2: Løse en andregradslikning

La oss se på hvordan vi kan løse likningen

$$
x^2 + x - 2 = 0,
$$

ved å bruke `sympy`. Vi kan løse denne likningen for hånd ved å faktorisere venstresiden av likningen, og da finner vi at $x = -2$ eller $x = 1$.

For å bruke `sympy.solve`, må vi identifiere hva $f(x)$ er her, og det er $f(x) = x^2 + x - 2$. Vi kan derfor skrive


In [43]:
import sympy
from sympy.abc import x

f = x**2 + x - 2 # Lager en symbolsk funksjon f(x) = x**2 + x - 2
løsninger = sympy.solve(f) # Løser likningen x**2 + x - 2 = 0.

print(løsninger)

[-2, 1]


som betyr at løsningen av likningen er $x = -2$ eller $x = 1$.

### Eksempel 3: Løse en tredjegradslikning

La oss se på et eksempel der vi skal løse en tredjegradslikning. Vi skal løse likningen

$$
x^3 + 2x^2 = x - 2.
$$

Første steg er å flytte alt over på venstresiden, slik at vi får en likning på formen $f(x) = 0$. Da får vi

$$
x^3 - 2x^2 - x + 2 = 0,
$$

så vi kan sette $f(x) = x^3 - 2x^2 - x + 2$ i tråd med det vi har gjort tidligere. For å løse likningen skriver vi derfor koden:

In [42]:
import sympy
from sympy.abc import x # Importerer symbolet for `x`

f = x**3 - 2 * x**2 - x + 2 # Lager en symbolsk funksjon f(x) = x**3 - 2*x**2 - x + 2
løsninger = sympy.solve(f) # Løser likningen x**3 - 2*x**2 - x + 2 = 0.

print(løsninger)

[-1, 1, 2]


som betyr at løsningen av likningen er $x = -1$ eller $x = 1$ eller $x = 2$.

#### Eksempel 4: Likning med brøk

La oss se på et eksempel der vi skal løse en likning med brøk. Vi skal løse likningen

$$
\frac{x^2 - x - 6}{x + 2} = 0.
$$

Vi kan løse likningen med `sympy` slik:

In [54]:
import sympy
from sympy.abc import x

f = (x**2 - x - 6) / (x + 2)
løsninger = sympy.solve(f) # Løser likningen (x**2 - x - 6) / (x + 2) = 0

print(løsninger)

[3]


som betyr at den eneste gyldige løsningen av likningen er $x = 3$.

```{admonition} Hva mener du med den *eneste gyldige* løsningen her?
:class: tip, dropdown

Her viser det seg at $x^2 - x - 6 = (x - 3)(x + 2)$, som betyr at $x = -2$ er en rot i uttrykket i telleren. Men dette er også en rot i nevneren. Dette er derfor ikke en gyldig løsning av likningen, fordi vi da deler på null. 

Heldigvis, trenger vi ikke bekymre oss for dette. `sympy` sjekker løsningene for oss og fjerner alle som leder til at vi deler på 0. 

```

### Eksempel 5: Likning med transcendentale funksjoner

La oss se på et eksempel der vi ønsker å finne nullpunktet til likningen

$$
e^x + 2x = 0.
$$

Her finnes det ingen analytisk løsning. Noen ganger uttrykkes svaret ved hjelp av spesielle funksjoner som vi ikke kjenner til i videregående skole. Som eksempel under:

In [46]:
import sympy
from sympy.abc import x

løsninger = sympy.solve(sp.exp(x) + 2*x)  # Løser likningen exp(x) + 2*x = 0.

print(løsninger)


[-LambertW(1/2)]


Dette er riktig svar, men forteller oss ikke stort. Vi kan få en numerisk verdi ved å bytte ut `sympy.solve` med `sympy.nsolve`, men da må oppgi en startverdi som vi tror er i nærheten av nullpunktet. Vi kan for eksempel prøve med $x_0 = -1$:

In [47]:
import sympy as sp
from sympy.abc import x

løsninger = sympy.nsolve(sp.exp(x) + 2*x, -1)  # Løser likningen exp(x) + 2*x = 0. Starter fra -1 som utgangspunkt.

print(løsninger)

-0.351733711249196


Altså er løsningen av likningen $x \approx -0.352$ med tre desimalers presisjon.

Nå fikk vi et tallsvar vi kan tolke! 

## Øvingsoppgaver

### Oppgave 1


Bruk `sympy` til å løse likningen

$$
3x - 2 = 0.
$$

*Under ligger et kodeskall du kan ta utgangspunkt i. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import sympy
from sympy.abc import x

f = NotImplemented # Skriv venstre side av likningen f(x) = 0
løsninger = NotImplemented # Løs likningen med sympy.solve

print(løsninger)

````{dropdown} Løsningsforslag

```python
import sympy
from sympy.abc import x

f = 3 * x - 2
løsninger = sympy.solve(3 * x - 2)

print(løsninger)
```

som gir utskriften

```console
[2/3]
```

Løsningen av likningen er altså $x = 2/3$

````

### Oppgave 2

Bruk `sympy` til å løse likningen

$$
x^2 - 3x + 2 = 0.
$$

*Under ligger et kodeskall du kan ta utgangspunkt i. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import sympy
from sympy.abc import x

f = NotImplemented # Skriv venstre side av likningen f(x) = 0
løsninger = NotImplemented # Løs likningen med sympy.solve

print(løsninger)

In [53]:
import sympy
from sympy.abc import x

f = x**2 - 3 * x + 2 
løsninger = sympy.solve(f)

print(løsninger)

[1, 2]


````{dropdown} Løsningsforslag

```python
import sympy
from sympy.abc import x

f = x**2 - 3 * x + 2 
løsninger = sympy.solve(f)

print(løsninger)
```

som gir utskriften

```console
[1, 2]
```

som betyr at løsningene er $x = 1$ eller $x = 2$.

````

### Oppgave 3 

Bruk `sympy` til å løse likningen

$$
x^3 - 13x^2 + 52x - 60 = 0.
$$

*Under ligger et kodeskall du kan ta utgangspunkt i. Du må fylle inn der det står `NotImplemented`. Her må du importere symbolet for `x` selv.*

In [ ]:
import sympy

# Her mangler importering av symbolet for `x`. Lag et symbol for `x` og løs likningen.

f = NotImplemented # Skriv venstre side av likningen f(x) = 0
løsninger = NotImplemented # Løs likningen

print(løsninger)

````{dropdown} Løsningsforslag

```python
import sympy
from sympy.abc import x

f = x**3 - 13 * x**2 + 52 * x - 60
løsninger = sympy.solve(f)

print(løsninger)
```

som gir utskriften

```console
[2, 5, 6]
```

som betyr at løsningene er $x = 2$ eller $x = 5$ eller $x = 6$.


````

### Oppgave 4

Bruk `sympy` til å løse likningen

$$
\frac{x^2 + 2x - 3}{x - 1} = 0.
$$ 

*Under ligger et kodeskall du kan ta utgangspunkt i. Du må fylle inn der det står `NotImplemented`. Her må du importere alt du trenger selv!*

In [ ]:
# Importerer det du trenger her

f = NotImplemented # Sett inn funksjonen her
løsninger = NotImplemented # Løs likningen her

print(løsninger)


````{dropdown} Løsningsforslag

```python
import sympy
from sympy.abc import x

f = (x**2 + 2 * x - 3) / (x - 1)
løsninger = sympy.solve(f)

print(løsninger)
```

som gir utskriften

```console
[-3]
```

som betyr at løsningen av likningen er $x = -3$. Her er $x = 1$ en rot i telleren og nevneren, som gjør at det ikke er en gyldig del av løsningen fordi vi ikke kan dele på null. `sympy` passer på å sjekke dette for oss slik at vi ikke trenger å tenke oss om!

````